kernels/singlecpu/src/actions/SCPUReactionImpls.cpp

```diff
@@ -364,6 +366,447 @@ void SCPUGillespie::perform(const util::PerformanceNode &node) {
+                    record->educts[1] = entry2.id;
+                    record->types_from[0] = entry1.type;
+                    record->types_from[1] = entry2.type;
+                }
+            }
+            break;
+        }
+        case readdy::model::actions::reactions::ConversionConversion: {
+            const auto &entry = data->entry_at(event.idx1);
+
+            // do not re-use entries for proper construction of backward update
+            readdy::model::Particle particle(entry.position(), reaction->products()[0]);
+            newParticles.emplace_back(particle);
+            decayedEntries.push_back(event.idx1);
+            energyDelta = 0.;
+
+            if(record) {
+                record->id = reaction->id();
+                record->type = static_cast<int>(reaction->type());
+                record->where = (entry.position());
+                bcs::fixPosition(record->where, box, pbc);
+                record->educts[0] = entry.id;
+                record->educts[1] = entry.id;
+                record->types_from[0] = entry.type;
+                record->types_from[1] = entry.type;
+            }
+
+            break;
+        }
+        case readdy::model::actions::reactions::EnzymaticEnzymatic: {
+            // find out which particle is the catalyst in A + C -> B + C
+            event_t::index_type catalystIdx;
+            event_t::index_type eductIdx;
+            if (event.t1 == reaction->educts()[1]) {
+                catalystIdx = event.idx1;
+                eductIdx = event.idx2;
+            } else if (event.t2 == reaction->educts()[1]){
+                catalystIdx = event.idx2;
+                eductIdx = event.idx1;
+            }
+
+            const auto& eductEntry = data->entry_at(eductIdx);
+            const auto& catalystEntry = data->entry_at(catalystIdx);
+
+            // do not re-use entries for proper construction of backward update
+            readdy::model::Particle particle(eductEntry.position(), reaction->products()[0]);
+            newParticles.emplace_back(particle);
+            decayedEntries.push_back(eductIdx);
+            energyDelta = 0.;
+
+            if(record) {
+                record->id = reaction->id();
+                record->type = static_cast<int>(reaction->type());
+                record->where = (eductEntry.position() + catalystEntry.position()) / 2.;
+                bcs::fixPosition(record->where, box, pbc);
+                record->educts[0] = eductEntry.id;
+                record->educts[1] = catalystEntry.id;
+                record->types_from[0] = eductEntry.type;
+                record->types_from[1] = catalystEntry.type;
+            }
+
+            break;
+        }
+        default: throw std::runtime_error("should not get here");
+    }
+
+    return std::make_pair(std::make_pair(std::move(newParticles), decayedEntries), energyDelta);
+}
+
+void SCPUDetailedBalance::calculateEnergies() {
+    const auto &context = kernel->context();
```

Comment:
this is more or less what the state model does to calculate the system's energy at a given step, can it be unified to avoid duplicate code? (eg through a template method that takes a function and evalutes that function on every particle that isn't deactivated and a function that does that for every interaction)

Reply:
Agreed